src/nxt_http_return.c

```diff
@@ -87,27 +97,123 @@ nxt_http_return(nxt_task_t *task, nxt_http_request_t *r,
         return NULL;
     }
 
+    ctx = nxt_mp_zget(r->mem_pool, sizeof(*ctx));
+    if (nxt_slow_path(ctx == NULL)) {
+        goto fail;
+    }
+
+    ctx->action = action;
     r->status = conf->status;
     r->resp.content_length_n = 0;
 
-    if (conf->location.length > 0) {
+    if (conf->loc_is_const) {
+        ctx->encoded = conf->encoded;
+
+        nxt_http_return_send_ready(task, r, ctx);
+
+    } else {
+        ret = nxt_var_query_init(&r->var_query, r, r->mem_pool);
+        if (nxt_slow_path(ret != NXT_OK)) {
+            goto fail;
+        }
+
+        nxt_var_query(task, r->var_query, conf->location, &ctx->location);
+
+        nxt_var_query_resolve(task, r->var_query, ctx,
+                              nxt_http_return_send_ready,
+                              nxt_http_return_var_error);
+    }
+
+    return NULL;
+
+fail:
+
+    nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+    return NULL;
+}
+
+
+static nxt_int_t
+nxt_http_return_encode(nxt_mp_t *mp, nxt_str_t *encoded,
+    const nxt_str_t *location)
+{
+    nxt_uint_t  encode;
+
+    if (nxt_is_complex_uri_encoded(location->start, location->length)) {
+        *encoded = *location;
+
+        return NXT_OK;
+    }
+
+    encode = nxt_encode_complex_uri(NULL, location->start, location->length);
+    encoded->length = location->length + encode * 2;
+
+    encoded->start = nxt_mp_nget(mp, encoded->length);
+    if (nxt_slow_path(encoded->start == NULL)) {
+        return NXT_ERROR;
+    }
+
+    nxt_encode_complex_uri(encoded->start, location->start, location->length);
+
+    return NXT_OK;
+}
+
+
+static void
+nxt_http_return_send_ready(nxt_task_t *task, void *obj, void *data)
+{
+    nxt_int_t               ret;
+    nxt_http_field_t        *field;
+    nxt_http_action_t       *action;
+    nxt_http_request_t      *r;
+    nxt_http_return_ctx_t   *ctx;
+    nxt_http_return_conf_t  *conf;
+
+    r = obj;
+    ctx = data;
+    action = ctx->action;
+    conf = action->u.conf;
+
+    if (!conf->loc_is_const) {
```

Comment:
What happens if the resolved variable value is empty?

Reply:
You mean this?:

```json
"action": {
    "return": 301,
    "location": "$host"
}
```
```
$ curl --dump-header - -H "Host: " localhost:82
HTTP/1.1 301 Moved Permanently
Server: Unit/1.27.0
Date: Thu, 07 Apr 2022 23:30:29 GMT
Content-Length: 0
```

The header will be skipped, as shown above.

Rationale:

http://www.alejandro-colomar.es/src/alx/nginx/unit.git/tree/src/nxt_http_return.c?h=location-var#n177

`$ find src -type f | xargs sed -n '/^nxt_http_return_send_ready/,/^}/p'`
```c
nxt_http_return_send_ready(nxt_task_t *task, void *obj, void *data)
{
    nxt_int_t               ret;
    nxt_http_field_t        *field;
    nxt_http_action_t       *action;
    nxt_http_request_t      *r;
    nxt_http_return_ctx_t   *ctx;
    nxt_http_return_conf_t  *conf;

    r = obj;
    ctx = data;
    action = ctx->action;
    conf = action->u.conf;

    // If there's a variable, it will enter this.
    if (!conf->loc_is_const) {
        // See below for details, but this just copies around in the case of an empty string.  No error.
        ret = nxt_http_return_encode(r->mem_pool, &ctx->encoded,
                                     &ctx->location);
        if (nxt_slow_path(ret == NXT_ERROR)) {
            goto fail;
        }
    }

    // Since the empty string has length 0, don't add a header.  Skip this.
    if (ctx->encoded.length > 0) {
        field = nxt_list_zero_add(r->resp.fields);
        if (nxt_slow_path(field == NULL)) {
            goto fail;
        }

        nxt_http_field_name_set(field, "Location");

        field->value = ctx->encoded.start;
        field->value_length = ctx->encoded.length;
    }


    // Basically it did nothing until here.  Only a few unused copies.
    r->state = &nxt_http_return_send_state;

    nxt_http_request_header_send(task, r, NULL, NULL);

    return;

fail:

    nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
}
```

`$ find src -type f | xargs sed -n '/^nxt_http_return_encode/,/^}/p'`
```
nxt_http_return_encode(nxt_mp_t *mp, nxt_str_t *encoded,
    const nxt_str_t *location)
{
    nxt_uint_t  encode;

    // The following check will return true, so it will just copy the string structure and exit successfully.  See below for details.
    if (nxt_is_complex_uri_encoded(location->start, location->length)) {
        *encoded = *location;

        return NXT_OK;
    }

    encode = nxt_encode_complex_uri(NULL, location->start, location->length);
    encoded->length = location->length + encode * 2;

    encoded->start = nxt_mp_nget(mp, encoded->length);
    if (nxt_slow_path(encoded->start == NULL)) {
        return NXT_ERROR;
    }

    nxt_encode_complex_uri(encoded->start, location->start, location->length);

    return NXT_OK;
}
```

`$ find src -type f | xargs sed -n '/^nxt_is_complex_uri_encoded/,/^}/p'`
```
nxt_is_complex_uri_encoded(u_char *src, size_t length)
{
    u_char   *reserved, *end, ch;
    uint8_t  d0, d1;

    reserved = (u_char *) "?#\0";

    // If the string is empty, length == 0, and end == src, so skip this loop and return true.
    for (end = src + length; src < end; src++) {
        ch = *src;

        if (nxt_uri_escape[ch >> 5] & (1U << (ch & 0x1f))) {
            if (ch == '%') {
                if (end - src < 2) {
                    return 0;
                }

                d0 = nxt_hex2int[*++src];
                d1 = nxt_hex2int[*++src];

                if ((d0 | d1) >= 16) {
                    return 0;
                }

                continue;
            }

            if (ch == reserved[0]) {
                reserved++;
                continue;
            }

            if (ch == reserved[1]) {
                reserved += 2;
                continue;
            }

            return 0;
        }
    }

    return 1;
}
```

Reply:
It's possible that `ctx->location.length` is NULL.
For example: `"location": "$arg_foo"`. In the future, we will support var like `$arg_*`.
```
// If there's a variable, it will enter this.
    if (!conf->loc_is_const) {
        // See below for details, but this just copies around in the case of an empty string.  No error.
        ret = nxt_http_return_encode(r->mem_pool, &ctx->encoded,
                                     &ctx->location);
        if (nxt_slow_path(ret == NXT_ERROR)) {
            goto fail;
        }
    }
```

Reply:
> It's possible that `ctx->location.length` is 0. For example: `"location": "$arg_foo"`. In the future, we will support var like `$arg_*`.

If you mean that the content of the variable is empty, that's no problem.  It's supported by the code, as shown above.  It simply sends no header (keeping the old behavior of considering the following equivalent, which I think is incorrect; but I think we should fix that in a different patch set, not here):

```
"action": {
    "return": 301
}
```
```
"action": {
    "return": 301,
    "location": ""
}
```

Check this branch, where I have a prototype (in fact, it's already tested and works): <https://github.com/alejandro-colomar/unit/tree/location-empty>

Reply:
```
// If there's a variable, it will enter this.
    if (!conf->loc_is_const) {
        // See below for details, but this just copies around in the case of an empty string.  No error.
        ret = nxt_http_return_encode(r->mem_pool, &ctx->encoded,
                                     &ctx->location);
        if (nxt_slow_path(ret == NXT_ERROR)) {
            goto fail;
        }
    }
```
Just to note if `ctx->location.length` is 0, `nxt_http_return_encode()` doesn't have to be called.

Reply:
> Check this branch, where I have a prototype (in fact, it's already tested and works):

I'm a bit confused now, does this patch depend on `empty location` patch for now?

Reply:
I know.  I just didn't want to optimize that case for the following reasons:

- It's unlikely that a server wants to send an empty Location header (because where would it redirect?  to itself?).
- It would add a check in the code that doesn't improve readability, nor makes the common case faster (in fact, it makes the common case ~1 op slower).

Reply:
> > Check this branch, where I have a prototype (in fact, it's already tested and works):
> 
> I'm a bit confused now, does this patch depend on `empty location` patch for now?

No.  It's the other way around.  ``location-empty` depends on this, and that's the reason I haven't published it yet (I'll do when we merge this one).

Reply:
The current patch set (aka `location`, aka `location-var`) behaves the same for an unset Location as for an empty string (`""`) as for an empty variable (`"$arg_foo"`).  It doesn't send a Location header.

Because previously, the code behaved the same for an empty string (`""`) as for an unset Location, and I wanted to keep that (slightly incorrect) behavior.  It didn't send a Location header.

_After_ we merge this, I'll propose a new patch set called `location-empty` which will behave correctly (but will not optimize) for empty strings (before or after variable resolution), by sending an empty Location header.  I will add a changelog line of type bugfix.